src/webview/wkwebview/mod.rs

```diff
@@ -329,66 +336,240 @@ impl InnerWebView {
-            match (function)(url.to_str().to_string(), is_main_frame) {
-              true => (*handler).call((1,)),
-              false => (*handler).call((0,)),
-            };
+          if should_download {
+            let has_download_handler = this.get_ivar::<*mut c_void>("HasDownloadHandler");
+            if !has_download_handler.is_null() {
+              let has_download_handler = &mut *(*has_download_handler as *mut Box<bool>);
+              if **has_download_handler {
+                (*handler).call((2,));
+              }
+            }
           } else {
-            log::warn!("WebView instance is dropped! This navigation handler shouldn't be called.");
-            (*handler).call((1,));
+            let function = this.get_ivar::<*mut c_void>("function");
+            if !function.is_null() {
+              let function = &mut *(*function as *mut Box<dyn for<'s> Fn(String, bool) -> bool>);
+              match (function)(url.to_str().to_string(), is_main_frame) {
+                true => (*handler).call((1,)),
+                false => (*handler).call((0,)),
+              };
+            } else {
+              log::warn!(
+                "WebView instance is dropped! This navigation handler shouldn't be called."
+              );
+              (*handler).call((1,));
+            }
+          }
+        }
+      }
+
+      // Navigation handler
+      extern "C" fn navigation_policy_response(
+        this: &Object,
+        _: Sel,
+        _: id,
+        response: id,
+        handler: id,
+      ) {
+        unsafe {
+          let handler = handler as *mut block::Block<(NSInteger,), c_void>;
+          let can_show_mime_type: bool = msg_send![response, canShowMIMEType];
+
+          if !can_show_mime_type {
+            let has_download_handler = this.get_ivar::<*mut c_void>("HasDownloadHandler");
+            if !has_download_handler.is_null() {
+              let has_download_handler = &mut *(*has_download_handler as *mut Box<bool>);
+              if **has_download_handler {
+                (*handler).call((2,));
+                return;
+              }
+            }
           }
+
+          (*handler).call((1,));
         }
       }
 
-      let navigation_decide_policy_ptr =
-        if attributes.navigation_handler.is_some() || attributes.new_window_req_handler.is_some() {
-          let cls = match ClassDecl::new("UIViewController", class!(NSObject)) {
+      let (navigation_decide_policy_ptr, download_delegate) = if attributes
+        .navigation_handler
+        .is_some()
+        || attributes.new_window_req_handler.is_some()
+        || attributes.download_started_handler.is_some()
+      {
+        let cls = match ClassDecl::new("UIViewController", class!(NSObject)) {
+          Some(mut cls) => {
+            cls.add_ivar::<*mut c_void>("function");
+            cls.add_ivar::<*mut c_void>("HasDownloadHandler");
```

Comment:
Can we extract these two strings into some consts so it is easier to track/change in the future? Actually any ivar used more than one time should be in a const

Reply:
There are many more ivar like this. I think we should do this in another PR.